Deployment for the cloud version (#1638)
CloudFire-LLC/cloudfire: elixir/README.md

> # Welcome to Elixir-land!

Reviewer: ```suggestion
# Welcome to Elixir-land!

This README provides an overview for running and managing Firezone's Elixir-based control plane.
```

## Running Control Plane for local development

> You can use [Top-Level Docker Compose](../docker-compose.yml) and start any released Elixir
> app. `web` and `api` services are running application release that will be pretty much the same
> as the one we run in production, while `elixir` compose service runs raw Elixir code, without a release.
> 
> It means that you can run any Elixir code including Mix tasks using `elixir` service but you can't do that
> in `web`/`api` so easily, because Elixir strips a lot of tooling during compilation.

Reviewer: ```suggestion
You can use the [Top-Level Docker Compose](../docker-compose.yml) to start any services locally. The `web` and `api` compose services are built application releases that are pretty much the same as the ones we run in production, while the `elixir` compose service runs raw Elixir code, without a built release.

This means you'll want to use the `elixir` compose service to run Mix tasks and any Elixir code on-the-fly, but you can't do that
in `web`/`api` so easily because Elixir strips out Mix and other tooling [when building an application release](https://hexdocs.pm/mix/Mix.Tasks.Release.html).
```

`elixir` additionally caches `_build` and `node_modules` to speed up compilation time and syncs
`/apps`, `/config` and other folders with the host machine.

```bash
# Make sure to run this every time code in elixir/ changes,

> # docket doesn't do that for you!

Reviewer: ```suggestion
# docker doesn't do that for you!
```

iex(web@127.0.0.1)1>
```

From `iex` shell you can run any Elixir code, for example you can emulate a full flow using process messages,
just keep in mind that you need to run seeds before executing this example:

```elixir
[gateway | _rest_gateways] = Domain.Repo.all(Domain.Gateways.Gateway)
:ok = Domain.Gateways.connect_gateway(gateway)

[relay | _rest_relays] = Domain.Repo.all(Domain.Relays.Relay)
relay_secret = Domain.Crypto.rand_string()
:ok = Domain.Relays.connect_relay(relay, relay_secret)
```

Now if you connect and list resources there will be one online because there is a relay and gateway online.

Some of the functions require authorization, here is how you can obtain a subject:

```elixir
user_agent = "User-Agent: iOS/12.7 (iPhone) connlib/0.7.412"
remote_ip = {127, 0, 0, 1}

# For a client
{:ok, subject} = Domain.Auth.sign_in(client_token, user_agent, remote_ip)

# For an admin user
provider = Domain.Repo.get_by(Domain.Auth.Provider, adapter: :userpass)
identity = Domain.Repo.get_by(Domain.Auth.Identity, provider_id: provider.id, provider_identifier: "firezone@localhost")
subject = Domain.Auth.build_subject(identity, nil, user_agent, remote_ip)
```

> ## Connecting to a staging or production instances

Reviewer: ```suggestion
## Connecting to a staging or production instances

We use Google Cloud Platform for all our staging and production infrastructure. You'll need access to this env to perform the commands below; ping someone for access if needed.
```

> This a danger zone so first of all, ALWAYS make sure on which environment your code is running:

Reviewer: ```suggestion
This is a danger zone so first of all, ALWAYS make sure on which environment your code is running:
```
